plonky3/src/circuit_builder.rs

```diff
@@ -154,18 +198,33 @@ impl<'a, T: FieldElement> PowdrCircuit<'a, T> {
 impl<'a, T: FieldElement> BaseAir<Val> for PowdrCircuit<'a, T> {
     fn width(&self) -> usize {
         assert_eq!(self.analyzed.constant_count(), 0);
-        self.analyzed.commitment_count()
+        self.analyzed.commitment_count() + self.analyzed.publics_count()
     }
 
     fn preprocessed_trace(&self) -> Option<RowMajorMatrix<Val>> {
         panic!()
     }
 }
 
-impl<'a, T: FieldElement, AB: AirBuilder<F = Val>> Air<AB> for PowdrCircuit<'a, T> {
+impl<'a, T: FieldElement, AB: AirBuilderWithPublicValues<F = Val>> Air<AB> for PowdrCircuit<'a, T> {
     fn eval(&self, builder: &mut AB) {
         let matrix = builder.main();
+        let pi = builder.public_values();
+
+        // public constraints
+        let pi_moved = pi.into_iter()
+            .map(|itm| *itm)
+            .collect::<Vec<<AB as AirBuilderWithPublicValues>::PublicVar>>();
```

Comment:
I don't think we need to move, we can access `pi[pub_idx]` lower down

Reply:
I think assertions use a mutable borrow of builder and pi are an immutable borrow-- I'm not sure how to do this without storing it in an intermediate